src/Migrator.sol

```diff
@@ -83,44 +83,101 @@ contract Migrator is IMigrator, Initializable, Ownable {
         swapSlippage = _amount;
     }
 
+    /// @inheritdoc IMigrator
+    function setApprovals() public onlyOwner {
+        ERC20(address(weth)).safeApprove(address(balancerVault), type(uint256).max);
+        token.safeApprove(address(balancerVault), type(uint256).max);
+        bpt.safeApprove(address(auraPool), type(uint256).max);
+        ERC20(address(slp)).safeApprove(address(sushiRouter), type(uint256).max);
+    }
+
     /*
-        Public functions
+        External functions
     */
 
     /// @inheritdoc IMigrator
-    function calculateSlpAmounts(uint256 _slpAmount) public view returns (uint256, uint256) {
+    function migrate(bool _stakeBpt) external {
+        uint256 slpBalance = slp.balanceOf(msg.sender);
+
+        ERC20(address(slp)).safeTransferFrom(msg.sender, address(this), slpBalance);
+
+        _unwrapSlp();
+
+        _swapWethForTokenBalancer();
+
+        _depositIntoBalancerPool();
+
+        // msg.sender receives auraBPT or BPT depending on their choice to deposit into Aura pool
+        if (_stakeBpt) _depositIntoRewardsPool();
+        else bpt.safeTransferFrom(address(this), msg.sender, bpt.balanceOf(address(this)));
+
+        emit Migrated(msg.sender, slpBalance, _stakeBpt);
+    }
+
+    /// @inheritdoc IMigrator
+    function userDepositIntoRewardsPool() external {
+        uint256 amount = bpt.balanceOf(msg.sender);
+        bpt.safeTransferFrom(msg.sender, address(this), amount);
+        auraPool.deposit(amount, msg.sender);
+    }
+
+    receive() external payable {
+        if (msg.value > 0) {
+            weth.deposit{ value: msg.value }();
+        }
+    }
+
+    /*
+        Internal functions
+    */
+
+    /**
+     * @notice Calculate the min amount of TOKEN and WETH for a given SLP amount
+     * @param _slpAmount The amount of SLP
+     * @return Return values for min amount out of TOKEN and WETH with unwrap slippage
+     */
+    function _calculateSlpAmounts(uint256 _slpAmount) internal view returns (uint256, uint256) {
+        uint256 tokenPriceEth = _tokenPrice();
+        uint256 wethPriceUsd = _wethPrice();
         uint256 slpSupply = slp.totalSupply();
         (uint256 wethReserves, uint256 tokenReserves, ) = slp.getReserves();
```

Comment:
Even when using an oracle, using the reserves in a calculation will result in potentially unsafe behavior. This calculation needs to be refactored. The only way to justify this would be to calculate the _expected_ reserves using the LP token reserves calculation.

Reply:
updated to move calculations off chain. One caveat is that the swap min amount out calculation is dependent on the result of unwrapping the SLP.